core/src/test/java/com/linecorp/armeria/server/websocket/WebSocketServiceTest.java

```diff
@@ -0,0 +1,201 @@
+                             .add(HttpHeaderNames.HOST, "foo.com")
+                             .add(HttpHeaderNames.ORIGIN, "http://foo.com")
+                             .addInt(HttpHeaderNames.SEC_WEBSOCKET_VERSION, 13)
+                             .add(HttpHeaderNames.SEC_WEBSOCKET_KEY, "dGhlIHNhbXBsZSBub25jZQ==")
+                             .add(HttpHeaderNames.SEC_WEBSOCKET_PROTOCOL, "superchat")
+                             .build();
+    }
+
+    static void checkCloseFrame(HttpData httpData, WebSocketCloseStatus closeStatus) {
+        // 0 ~ 3 FIN, RSV1, RSV2, RSV3. 4 ~ 7 opcode
+        final ByteBuf byteBuf = httpData.byteBuf();
+        assertThat(byteBuf.readByte() & 0x0F).isEqualTo(WebSocketFrameType.CLOSE.opcode());
+        // Skip 1 bytes.
+        byteBuf.readByte();
+        assertThat((int) byteBuf.readShort()).isEqualTo(closeStatus.code());
+    }
+
+    @Test
+    void testDecodedContinuationFrame() throws Exception {
+        final CompletableFuture<List<WebSocketFrame>> collectFuture = new CompletableFuture<>();
+        final WebSocketService webSocketService = WebSocketService.of((ctx, in) -> {
+            in.collect().thenAccept(collectFuture::complete);
+            return WebSocket.streaming();
+        });
+
+        final HttpResponse response = webSocketService.serve(ctx, req);
+        req.write(toHttpData(WebSocketFrame.ofText("foo", false)));
+        req.write(toHttpData(WebSocketFrame.ofContinuation("bar", true)));
+        req.write(toHttpData(WebSocketFrame.ofBinary("foo".getBytes(StandardCharsets.UTF_8), false)));
+        req.write(toHttpData(WebSocketFrame.ofContinuation("bar".getBytes(StandardCharsets.UTF_8), true)));
+        req.close();
+
+        response.subscribe(NoopSubscriber.get());
+        final List<WebSocketFrame> frames = collectFuture.join();
+        assertThat(frames.size()).isEqualTo(4);
+        WebSocketFrame frame = frames.get(0);
+        assertThat(frame.isFinalFragment()).isFalse();
+        assertThat(frame.type()).isSameAs(WebSocketFrameType.TEXT);
+        assertThat(frame.text()).isEqualTo("foo");
+
+        frame = frames.get(1);
+        assertThat(frame.isFinalFragment()).isTrue();
+        assertThat(frame.type()).isSameAs(WebSocketFrameType.CONTINUATION);
+        assertThat(frame.text()).isEqualTo("bar");
+
+        frame = frames.get(2);
+        assertThat(frame.isFinalFragment()).isFalse();
+        assertThat(frame.type()).isSameAs(WebSocketFrameType.BINARY);
+        assertThat(frame.text()).isEqualTo("foo");
+
+        frame = frames.get(3);
+        assertThat(frame.isFinalFragment()).isTrue();
+        assertThat(frame.type()).isSameAs(WebSocketFrameType.CONTINUATION);
+        assertThat(frame.text()).isEqualTo("bar");
+    }
+
+    private HttpData toHttpData(WebSocketFrame frame) {
+        return HttpData.wrap(encoder.encode(ctx, frame));
+    }
+
+    static class AbstractWebSocketHandler implements WebSocketHandler {
```

Comment:
We could add something similar in our public API.

Reply:
That's in the todo list. 😉